crates/query-engine/translation/src/translation/query/filtering.rs

```diff
@@ -60,39 +60,41 @@ pub fn translate_expression_with_joins(
     match predicate {
         models::Expression::And { expressions } => {
             let mut acc_joins = vec![];
-            let and_exprs = expressions
-                .iter()
-                .map(|expr| translate_expression_with_joins(env, state, current_table_scope, expr))
-                .try_fold(
-                    sql::ast::Expression::Value(sql::ast::Value::Bool(true)),
-                    |acc, expr| {
-                        let (right, right_joins) = expr?;
-                        acc_joins.extend(right_joins);
-                        Ok(sql::ast::Expression::And {
-                            left: Box::new(acc),
-                            right: Box::new(right),
-                        })
-                    },
-                )?;
-            Ok((and_exprs, acc_joins))
+            let mut and_expressions = Vec::new();
+
+            // Translate all AND expressions and collect joins
+            for expr in expressions {
+                let (translated_expr, expr_joins) =
+                    translate_expression_with_joins(env, state, current_table_scope, expr)?;
+                acc_joins.extend(expr_joins);
+                and_expressions.push(translated_expr);
+            }
+
+            // Create a balanced AND tree to avoid stack overflow
+            let final_expr = sql::helpers::fold_and_balanced(and_expressions);
+            Ok((final_expr, acc_joins))
         }
         models::Expression::Or { expressions } => {
+            // Try to optimize OR expressions to IN operations first
+            if let Some((optimized_expr, optimized_joins)) =
+                try_optimize_or_to_in(env, state, current_table_scope, expressions)? {
```

Comment:
This other optimization seems nice, but quite magic, let's move this to a follow-up PR in case we need to bisect a production problem later.